third_party/amd/lib/TritonAMDGPUToLLVM/UpcastMXFPToLLVM.cpp

```diff
@@ -234,20 +230,166 @@ Value mxfpScaleBf16ViaF32(RewriterBase &rewriter, Location loc, Value v,
+  // Intentionally swap the byte indices 1 and 2 to align with how the LLVM
+  // backend accesses them
+  results.push_back(rewriter.create<ConvertOp>(loc, resType, packedVec,
+                                               scaleF32,
+                                               /*srcSelIndex=*/2));
+  results.push_back(rewriter.create<ConvertOp>(loc, resType, packedVec,
+                                               scaleF32,
+                                               /*srcSelIndex=*/1));
+  results.push_back(rewriter.create<ConvertOp>(loc, resType, packedVec,
+                                               scaleF32,
+                                               /*srcSelIndex=*/3));
+  return results;
+}
+
+template <typename ConvertOp>
+SmallVector<Value, 2> upcast4xMxfp8_HW(RewriterBase &rewriter, Location loc,
+                                       const SmallVector<Value> &xVals, int idx,
+                                       Value scale) {
+  auto b = TritonLLVMOpBuilder(loc, rewriter);
+  Value v0 = xVals[idx];
+  Value v1 = xVals[idx + 1];
+  Value v2 = xVals[idx + 2];
+  Value v3 = xVals[idx + 3];
+  Value packedVec = b.undef(vec_ty(i8_ty, 4));
+  packedVec = b.insert_element(packedVec, v0, b.i32_val(0));
+  packedVec = b.insert_element(packedVec, v1, b.i32_val(1));
+  packedVec = b.insert_element(packedVec, v2, b.i32_val(2));
+  packedVec = b.insert_element(packedVec, v3, b.i32_val(3));
+  packedVec = b.bitcast(packedVec, i32_ty);
+  Type retElemType = bf16_ty;
+  if constexpr (std::is_same_v<ConvertOp, ROCDL::CvtScaleF32PkF16Fp8Op> ||
+                std::is_same_v<ConvertOp, ROCDL::CvtScaleF32PkF16Bf8Op>)
+    retElemType = f16_ty;
+  Type resType = vec_ty(retElemType, 2);
+  Value scaleF32 =
+      b.bitcast(b.shl(b.zext(i32_ty, scale), b.i32_val(23)), f32_ty);
+  SmallVector<Value, 2> results;
+  results.push_back(rewriter.create<ConvertOp>(loc, resType, packedVec,
+                                               scaleF32,
+                                               /*srcLoHiSel=*/false));
+  results.push_back(rewriter.create<ConvertOp>(loc, resType, packedVec,
+                                               scaleF32,
+                                               /*srcLoHiSel=*/true));
+  return results;
+}
+
+// Upcast 8 mxfp4 values from xVals starting at idx using the given scale
+// factor, and store the results into yVals
+static void upcast8xMxfp4(RewriterBase &rewriter, Location loc,
+                          AMD::ISAFamily isaFamily, amdgpu::UpcastMXFPOp op,
+                          const SmallVector<Value> &xVals, bool useFp16,
+                          int idx, Value scale, SmallVector<Value> &yVals) {
+  auto b = TritonLLVMOpBuilder(loc, rewriter);
+  if (isaFamily == AMD::ISAFamily::CDNA4) {
+    Type retElemType = useFp16 ? f16_ty : bf16_ty;
+    SmallVector<Value, 4> v4i32 =
+        useFp16 ? upcast8xMxfp4_HW<ROCDL::CvtScaleF32PkF16Fp4Op>(
```

Comment:
Maybe additionally test [here](https://github.com/triton-lang/triton/blob/cf0db926f29837df6b36eae3c01e12cfae08528a/python/test/unit/language/test_core.py#L4436) that indeed we generate such `V_CVT_SCALEF32_PK32_*` instructions in amdgcn for cdna4?